File: java/beer-song/src/main/java/BeerSong.java
public class BeerSong {

    private final static String PHRASE_1 = "%d bottle%s of beer on the wall, %d bottle%s of beer."
            + System.getProperty("line.separator");
    private final static String PHRASE_2 = "Take %s down and pass it around, %d bottle%s of beer on the wall."
            + System.getProperty("line.separator") + System.getProperty("line.separator");
    private final static String PHRASE_ONE_BOTTLE = "Take it down and pass it around, no more bottles of beer on the wall."
            + System.getProperty("line.separator") + System.getProperty("line.separator");
    private final static String PHRASE_ZERO_BOTTLE_1 = "No more bottles of beer on the wall, no more bottles of beer."
            + System.getProperty("line.separator");
    private final static String PHRASE_ZERO_BOTTLE_2 = "Go to the store and buy some more, 99 bottles of beer on the wall."
            + System.getProperty("line.separator") + System.getProperty("line.separator");

    public String sing(int bottles, int verses) {
        StringBuilder sb = new StringBuilder();

        // This is a signal to end recursion
        if (verses == 0) {
            return "";
        }

        if (bottles == 0) {
            // Last two sentences of the song aren't reusable
            return sb.append(PHRASE_ZERO_BOTTLE_1).append(PHRASE_ZERO_BOTTLE_2).toString();
        }

        // Note: after this, we'll invoke again sing method in the return statements: recursion to build the song

        if (bottles == 1) {
            // Take care of singular and no more bottles
            sb.append(String.format(PHRASE_1, bottles, "", bottles, ""));
            return sb.append(PHRASE_ONE_BOTTLE).append(sing(bottles - 1, verses - 1)).toString();
        }

        sb.append(String.format(PHRASE_1, bottles, "s", bottles, "s"));

        if (bottles == 2) {
            // Take care of singular in the second phrase
            return sb.append(String.format(PHRASE_2, "one", bottles - 1, ""))
                    .append(sing(bottles - 1, verses - 1)).toString();
        }

        return sb.append(String.format(PHRASE_2, "one", bottles - 1, "s"))
                .append(sing(bottles - 1, verses - 1)).toString();

    }

    public String singSong() {
        return sing(99, 100);
    }
}

File: scala/beer-song/src/main/scala/BeerSong.scala
object BeerSong {
  def reciteVerse(verse: Int): String = verse match {
    case 0 =>
      "No more bottles of beer on the wall, no more bottles of beer.\nGo to the store and buy some more, 99 bottles of beer on the wall.\n"
    case 1 =>
      "1 bottle of beer on the wall, 1 bottle of beer.\nTake it down and pass it around, no more bottles of beer on the wall.\n"
    case 2 =>
      "2 bottles of beer on the wall, 2 bottles of beer.\nTake one down and pass it around, 1 bottle of beer on the wall.\n"
    case _ => s"$verse bottles of beer on the wall, $verse bottles of beer.\n" +
      s"Take one down and pass it around, ${verse - 1} bottles of beer on the wall.\n"
  }

  def recite(verse: Int, numberOfVerses: Int): String = {
    ((verse - numberOfVerses + 1) to verse).reverse.map(reciteVerse(_)).reduce(_ + "\n" + _)
  }
}
